design/spark-co-exist-with-object-storages.md

```diff
@@ -0,0 +1,141 @@
+# Spark: co-existing with existing underlying object store - Design Options
+
+This document includes a proposed solution for https://github.com/treeverse/lakeFS/issues/2625. 
+Its intent is to put an idea in writing (before even testing it), and not to suggest that this is the best design option. 
+To choose a solution, we need to do more research and consider alternatives that will hopefully add less amount of friction.
+
+## Goals
+
+### Must do
+
+1. Provide a configurable set of overrides to translate paths at runtime from an old object store location to a new lakeFS location.
+2. Support any object storage that lakeFS supports.
+3. Do something that adds the least amount of friction to a Spark user as possible.
+
+### Nice to have
+
+Make the solution usable by non-lakeFS users by:
+1. Supporting path translation from and to any type of path.
+2. Make the solution a standalone component with no direct dependency on anything other than Spark itself. 
+
+## Non-goals
+
+1. Convert users from [accessing lakeFS from S3 gateway](../docs/integrations/spark.md#access-lakefs-using-the-s3a-gateway) to [accessing lakeFS with lakeFS-specific Hadoop filesystem](../docs/integrations/spark.md#access-lakefs-using-the-lakefs-specific-hadoop-filesystem).  
+
+## Proposal: Introducing RouterFileSystem
+
+We would like to implement a [HadoopFileSystem](https://github.com/apache/hadoop/blob/2960d83c255a00a549f8809882cd3b73a6266b6d/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java) 
+that translates object store URIs into lakeFS URIs according to a configurable mapping, and uses a relevant Hadoop file system to 
+perform any file system operation.  
+
+### Handle any interaction with the underlying object store
+
+To allow Spark users to integrate with lakeFS without changing their Spark code, `RouterFileSystem` is configured to be 
+the file system for object store URIs with a certain scheme. For example, the following Spark configurations make 
+`RouterFileSystem` the file system for URIs with `scheme=s3`:
+```shell
+fs.s3.impl=RouterFileSystem
+```
+This will force any file system operation performed on an ibject URI with `scheme=s3` to go through `RouterFileSystem`.   
+
+### URI translation
+
+`RouterFileSystem` has access to a configurable mapping that maps bucket names into lakeFS repositories and prefixes, e.g.
+```yaml
+MapperConfig = {
+    {
+        "source_bucket_pattern": "bucket-a",
+        "mapped_bucket_name": "example-repo",
```

Comment:
maybe we should consider more "git-like" names? (e.g. "mapped-repository" and "mapped branch")